### PR TITLE
Changed schema and Added new Tables

### PR DESCRIPTION
**Pull request by pacospace**
- Changed schema to separate internal from external knowledge of Thoth  
- Changed sync logic to increase robustness
- Corrected bugs
- Standardized runtime environment syncing
- Created `is_external` for PackageExtractRun

Signed-off-by: Francesco Murdaca <fmurdaca@redhat.com>

### CODE REVIEW

File: thoth/storages/graph/postgres.py [hunk 1]
```diff
@@ -1645,63 +1648,141 @@ def _create_python_package_requirement(self, requirements: dict) -> List[PythonP
...
+                ))
 
-        return result
+            return result
+        else:
+            result = []
+            pipfile_locked = PipfileLock.from_dict(pipfile_locked, pipfile=None)
+            os_name = software_environment.os_name
+            os_version = software_environment.os_version
+            python_version = software_environment.python_version
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: These variables could be unrolled up as both branches use them.

File: thoth/storages/graph/postgres.py [hunk 2]
```diff
@@ -1645,63 +1648,141 @@ def _create_python_package_requirement(self, requirements: dict) -> List[PythonP
...
+                        .filter(PythonPackageVersion.package_version == package.locked_version)
+                        .filter(PythonPackageVersion.os_name == os_name)
+                        .filter(PythonPackageVersion.os_version == os_version)
+                        .filter(PythonPackageVersion.python_version == python_version)
+                        .filter(PythonPackageIndex.url == package.index.url if package.index else None)
+                        .first()
+                    )
+                else:
+                    logging.warning("Package %s %s does not have Index!!!" % (package.name, package.locked_version))
+                    raise Exception("Package %s %s does not have Index!!!" % (package.name, package.locked_version))
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: This exception is too generic please create a new one in exceptions.py file and derive from `ThothStorageException` so we can handle this specific exception in upper layers if needed.

Logging of warning before raising an exception can be removed.

And the last one - exclamation marks seem to be too imperative :), you can provide a more gentle error message (f-strings are allowed):
```python
f"Trying to sync package {package.name!r} in version {package.locked_version!r} which does not have corresponding Python entity record"
```

File: thoth/storages/graph/postgres.py [hunk 3]
```diff
@@ -1645,63 +1648,141 @@ def _create_python_package_requirement(self, requirements: dict) -> List[PythonP
...
+                    ))
+                else:
+                    logging.warning(
+                        "Solver for %s-%s-%s didn't run yet for %s %s!!!"
+                        % (os_name, os_version, python_version, package.name, package.locked_version)
+                        )
+                    raise Exception(
+                        "Solver for %s-%s-%s didn't run yet for %s %s!!!"
+                        % (os_name, os_version, python_version, package.name, package.locked_version)
+                        )
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: Same as for the exception above.

File: thoth/storages/graph/postgres.py [hunk 4]
```diff
@@ -1645,63 +1648,141 @@ def _create_python_package_requirement(self, requirements: dict) -> List[PythonP
...
+
+            return result
 
     def _runtime_environment_conf2models(
         self,
         runtime_environment: dict,
         environment_type: str,
-        is_user: bool,
+        is_user: bool
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: This can be renamed to `is_external`.

File: thoth/storages/graph/postgres.py [hunk 5]
```diff
@@ -2299,6 +2315,9 @@ def sync_analysis_result(self, document: dict) -> None:
         os_name = document["result"]["operating-system"]["name"]
         os_version = document["result"]["operating-system"]["version_id"]
 
+        # Check if it comes from a User
+        is_external = document["metadata"]["arguments"]["thoth-package-extract"]["metadata"]["is_external"]
```
Comment (fridex) [inline thoth/storages/graph/postgres.py]: To keep ability to sync older results and wait also for changes to be propagated to deployment, I would suggest:

```python
        is_external = document["metadata"]["arguments"]["thoth-package-extract"]["metadata"].get("is_external", True)
```

File: thoth/storages/graph/postgres.py [hunk 6]
```diff
@@ -2181,6 +2315,9 @@ def sync_analysis_result(self, document: dict) -> None:
...
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py]: Linked to: https://github.com/thoth-station/user-api/pull/593/files @fridex :)

File: thoth/storages/graph/postgres.py (hunk 5 above)
Comment (pacospace) [reply thoth/storages/graph/postgres.py]: Perfect, I will adjust it, Thanks :)

**Pull request merged**